src/clj/rems/db/dynamic_roles.clj

```diff
@@ -42,15 +42,22 @@
     (is (= #{}
            (roles-from-all-applications "unknown" events)))))
 
-(mount/defstate dynamic-roles-cache
+(mount/defstate
+  ^{:doc "The cached state will contain the following keys:
+          ::permissions
+          - Map from application ID to the roles and permissions of an application.
+          ::roles-by-user
+          - Map from user ID to a set of all application roles which the user has,
+            a union of roles from all applications."}
+  dynamic-roles-cache
   :start (events-cache/new))
 
 (defn get-roles [user]
   (->> (events-cache/refresh!
         dynamic-roles-cache
         (fn [state events]
-          (let [apps (reduce permissions-of-all-applications (::apps state) events)]
-            {::apps apps
-             ::roles-by-user (group-roles-by-user apps)})))
+          (let [permissions (reduce permissions-of-all-applications (::permissions state) events)]
+            {::permissions permissions
+             ::roles-by-user (group-roles-by-user permissions)})))
```

Comment:
is `group-roles-by-user` really so slow that we need to memoize it?

Reply:
There are benchmark measurements in the commit message 87a598bafa9d210b31d690e70d85f8932a370615. It used to be a linear search which calls `apply-user-permissions` for every application, but now it's just a hashmap lookup. Filtering a person's own applications and open/handled reviews is still a linear search, but most users can see only a few applications so it's less of an issue.